### PR TITLE
[KSM] Add basic gpu metrics

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR adds the following metrics:
- `kubernetes_state.node.gpu_capacity` tagged by `node`, `resource`, `unit` and `mig_profile`
- `kubernetes_state.node.gpu_allocatable` tagged by `node`, `resource`, `unit` and `mig_profile`
- `kubernetes_state.container.gpu_limit`  tagged by `kube_namespace`, `pod_name`,  `kube_container_name`, `node`, `resource`,  `unit`, `mig_profile`.

`resource` will be `amd_com_gpu`, `nvidia_com_gpu`, `gpu_intel_com_i915` or `nvidia_com_mig_2c_4g_20gb`.
`mig_profile` is tagged only for Nvidia Multi-Instance-GPU.
It was kept this way for code simplicity and to keep the vendor of the GPU.

### Motivation
We want to monitor GPU resources in Kubernetes.

### Additional Notes
Need to update the doc in integrations-core repo

### Possible Drawbacks / Trade-offs
N/A

### Describe how to test/QA your changes
Deploy an EKS cluster with GPU instances. Normally, `kubectl describe nodes` should show the GPU resources. Make sure the KSM core check emits the GPU metrics with the correct tags.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by estherk15 — APPROVED** [review]

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 1]
```diff
@@ -161,18 +168,34 @@ func (a *sumValuesAggregator) accumulate(metric ksmstore.DDMetric) {
...
+		return "gpu"
+	}
+	return resource
+}
+
 func (a *resourceAggregator) accumulate(metric ksmstore.DDMetric) {
+	resource := renameMig(metric.Labels["resource"])
+	if newName, ok := renamedResource[resource]; ok {
+		resource = newName
+	}
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:193]: Maybe we can move all the code related to resource renaming in a single function:
* move the `renamedResource` lookup inside the function
* rename the `renameMig` function to `renameResource`

```suggestion
var gpuResource = map[string]struct{}{
	"nvidia_com_gpu":     struct{}{},
	"amd_com_gpu":        struct{}{},
	"gpu_intel_com_i915": struct{}{},
}

func renameResource(resource string) string {
	if strings.HasPrefix(resource, "nvidia_com_mig") {
		return "gpu"
	}
	
	if _, found := renamedResource[resource]; found {
		return "gpu"
	}
	
	return resource
}

func (a *resourceAggregator) accumulate(metric ksmstore.DDMetric) {
	resource := renameResource(metric.Labels["resource"])
```

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 2]
```diff
@@ -234,12 +255,12 @@ func (a *counterAggregator) flush(sender sender.Sender, k *KSMCheck, labelJoiner
...
 		labels := make(map[string]string)
-		for i, allowedLabel := range a.allowedLabels {
-			if allowedLabel == "" {
+		for i := range a.allowedLabels {
+			if a.allowedLabels[i] == "" {
 				break
 			}
 
-			labels[allowedLabel] = labelValues[i]
+			labels[a.allowedLabels[i]] = labelValues[i]
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:263]: We can keep `allowedLabel` instead of `a.allowedLabels[i]`.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 3]
```diff
@@ -23,6 +23,42 @@ import (
...
+	"nvidia_com_gpu": "gpu",
+	// Note: i915 is a driver name. It is unsure if other drivers are currently supported by intel
+	"gpu_intel_com_i915": "gpu",
+}
+
+// resourceDDName returns the datadog name of the given resource with possibly extra tags
+func resourceDDName(resource string) (string, []string, bool) {
+	// Handle most resources
+	if ddname, allowed := allowedResources[resource]; allowed {
+		return ddname, nil, allowed
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:47]:
```suggestion
		return ddname, nil, true
```

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 4]
```diff
@@ -161,18 +168,34 @@ func (a *sumValuesAggregator) accumulate(metric ksmstore.DDMetric) {
...
 
-	for i, allowedLabel := range a.allowedLabels {
-		if allowedLabel == "" {
+	for i := range a.allowedLabels {
+		if a.allowedLabels[i] == "" {
 			break
 		}
 
-		labelValues[i] = metric.Labels[allowedLabel]
+		labelValues[i] = metric.Labels[a.allowedLabels[i]]
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:176]: KSM uses a lot of memory and using `a.allowedLabels[i]` copies the entire string. However, considering that this code is not called very often, I guess we can favor readability

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 5]
```diff
@@ -147,12 +154,12 @@ func newLastCronJobAggregator() *lastCronJobAggregator {
...
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go]: It's surprising that it's faster to copy the string. I was more worried about the memory usage involved by copying the strings. I do agree that considering this code is not called so often, it might be better to favor readability.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 6]
```diff
@@ -228,27 +264,18 @@ func containerResourceLimitsTransformer(s sender.Sender, name string, metric ksm
 // submitContainerResourceMetric can be called by container resource metric transformers to submit resource-specific metrics
 // metricSuffix can be either requested or limit
 func submitContainerResourceMetric(s sender.Sender, name string, metric ksmstore.DDMetric, hostname string, tags []string, metricSuffix string) {
-
-	var allowedResources = map[string]string{
```
Comment (AliDatadog) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:232]: @L3n41c We merged `allowedResources` for containers and nodes. I wonder if I should revert this change as this could add weird resources.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 7]
```diff
@@ -23,6 +23,68 @@ import (
...
+	return resource
+}
+
+func resourceExtraTag(resource string) []string {
+	if strings.HasPrefix(resource, "nvidia_com_mig") {
+		splitStr := strings.Split(resource, "mig_")
+		if len(splitStr) == 2 {
+			return []string{fmt.Sprintf("mig_profile:%s", strings.Replace(splitStr[1], "_", "-", 2))}
+		}
+	}
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:70]: The use of `strings.Split(…)` looks a little bit overkill as, thanks to the `if strings.HasPrefix(…)`, we already know what the first token will be.
Basically, if we matched the `nvidia_com_mig` prefix, what we need is just what is after.
Also, is it on purpose that we replace only the first two `_` in `-` ?
I looked at the tests and couldn’t find a case where we would be interested in keeping a potential third `_` as is.

```suggestion
	if suffix, found := strings.CutPrefix(resource, "nvidia_com_mig_"); found {
		return []string{"mig_profile:" + strings.ReplaceAll(suffix, "_", "-")}
	}
```

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go [hunk 8]
```diff
@@ -23,6 +23,68 @@ import (
...
+	extraTag := resourceExtraTag(resource)
+
+	// Rename the resource
+	ddname := renameResource(resource)
+
+	// Check if the renamed resource is allowed
+	_, allowed := allowedResources[ddname]
+
+	return ddname, extraTag, allowed
+}
```
Comment (L3n41c) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go:86]:
```suggestion
// resourceDDName returns the datadog name of the given resource with possibly extra tags.
func resourceDDName(resource string, allowedResources map[string]struct{}) (ddname string, extraTag []string, allowed bool) {
	// Add an extra tag
	extraTag = resourceExtraTag(resource)

	// Rename the resource
	ddname = renameResource(resource)

	// Check if the renamed resource is allowed
	_, allowed = allowedResources[ddname]

	return
}
```

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go (hunk 7 above)
Comment (AliDatadog) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go]: Good catch. `CutPrefix` is indeed what I wanted to build. I replaced only the first two "_" because normally there shouldn't be a third `_` but indeed we can replace all of them.

**Review by L3n41c — APPROVED** [review]
